### PR TITLE
merging backend logic

### PR DESCRIPTION
**Pull request by amjedidiah**

### CODE REVIEW

File: src/frontend/lib/uploadObject.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+            Object.entries(fields).forEach(([key, value]) => {
+                formData.append(key, value)
+            })
+        formData.append('file', file)
+
+        const response = await fetch(presignedUrl, {
+            method: 'PUT',
+            body: fields ? formData : file,
+            headers: {
+                'Content-Type': file.type,
```
Comment (BSalaeddin) [inline src/frontend/lib/uploadObject.ts]: When sending FormData, manually setting the 'Content-Type' header interferes with the automatic setting of the boundary parameter, causing the request to fail.

Recommendation: Remove the 'Content-Type' header when fields are present and FormData is used. Let the browser set it automatically. Modify the code as follows:

```
const response = await fetch(presignedUrl, {
    method: 'PUT',
    body: fields ? formData : file,
    headers: fields
        ? undefined // Do not set 'Content-Type' header when sending FormData
        : {
              'Content-Type': file.type,
          },
    mode: 'cors',
    credentials: 'include',
})

```

File: src/frontend/lib/storage/provider.ts [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+        if (missing.length > 0)
+            throw new Error(
+                `Missing required configuration: ${missing.join(', ')}`,
+            )
+
+        return true
+    }
+
+    private handleError(error: unknown): never {
```
Comment (BSalaeddin) [inline src/frontend/lib/storage/provider.ts:184]: Issue: The method relies on checking substrings in err.message to determine the error type. This approach is fragile and can break if error messages change or are localized.

Recommendation: Use more structured error handling by checking specific error properties like name, code, or custom properties returned from the backend. For example, if the backend returns an error object with a code property, you can use that to determine the error type:


private handleError(error: unknown): never {
    if (error instanceof UploadError) throw error;

    const err = error as any;
    if (err.code === 'UNAUTHORIZED') {
        throw new UploadError(
            'Unauthorized access to Provider',
            UploadErrorType.PERMISSION_ERROR
        );
    }

    if (err.code === 'EXPIRED_URL') {
        throw new UploadError(
            'Presigned URL has expired',
            UploadErrorType.EXPIRED_URL,
            true
        );
    }

    throw new UploadError(
        `Upload failed: ${err.message}`,
        UploadErrorType.UNKNOWN_UPLOAD_ERROR,
        true
    );
}

File: src/frontend/lib/storage/provider.ts [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+            )
+
+            xhr.open('PUT', url)
+            if (this.config.provider === Provider.Azure)
+                xhr.setRequestHeader('x-ms-blob-type', 'BlockBlob')
+            xhr.send(file)
+        })
+    }
+
+    validateConfig(): boolean {
```
Comment (BSalaeddin) [inline src/frontend/lib/storage/provider.ts:170]: Validation of Required Configurations:

Issue: The validateConfig method only checks for tokenEndpoint, but provider is also critical for the functionality.

Recommendation: Update the validateConfig method to include provider:

validateConfig(): boolean {
    const required = ['tokenEndpoint', 'provider'] as const;
    const missing = required.filter(key => !this.config[key]);

    if (missing.length > 0) {
        throw new Error(`Missing required configuration: ${missing.join(', ')}`);
    }

    // Optionally, validate that provider is a valid enum value
    if (!Object.values(Provider).includes(this.config.provider)) {
        throw new Error(`Invalid provider: ${this.config.provider}`);
    }

    return true;
}

File: src/frontend/lib/storage/provider.ts [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+                    reject(
+                        new Error(
+                            `Upload failed with status ${xhr.status}: ${xhr.responseText}`,
+                        ),
+                    )
+                }
+            })
+
+            xhr.addEventListener('error', () =>
+                reject(new Error('Network error during upload')),
```
Comment (BSalaeddin) [inline src/frontend/lib/storage/provider.ts]: Recommendation: Enhance error handling by including xhr.status and xhr.statusText:

```
xhr.addEventListener('error', () =>
    reject(
        new Error(
            `Network error during upload: ${xhr.status} ${xhr.statusText}`,
        ),
    ),
);

```

File: src/frontend/lib/storage/provider.ts (hunk 3 above)
Comment (amjedidiah) [reply src/frontend/lib/storage/provider.ts:170]: Thanks @BSalaeddin  done

File: src/frontend/lib/storage/provider.ts (hunk 4 above)
Comment (amjedidiah) [reply src/frontend/lib/storage/provider.ts]: Thanks @BSalaeddin , done

File: src/frontend/lib/storage/provider.ts (hunk 2 above)
Comment (amjedidiah) [reply src/frontend/lib/storage/provider.ts:184]: Thanks @BSalaeddin , done

File: src/frontend/lib/uploadObject.ts (hunk 1 above)
Comment (amjedidiah) [reply src/frontend/lib/uploadObject.ts]: Thanks @BSalaeddin , done

**Review by BSalaeddin — APPROVED** [review]

**Review by MedAmine1212 — APPROVED** [review]

**Pull request merged**